src/amp-story-player/amp-story-player-iframe-pool.js

```diff
@@ -77,31 +90,145 @@ export class IframePool {
+   *     |                            v                           |
+   *     |               +-------------------------------         |
+   *     |               |                             |          |
+   *     |               | mark iframe to load its src |        N |
+   *     |               |                             |          |
+   *     |               +-----------------------------+          |
+   *     |                            |                           |
+   *     |                            |                           |
+   *     |                            v                           |
+   *     |                +---------------------------+           |
+   *     |                |                           |           |
+   *     |                |   mark its positioning    |           |
+   *     |                |(previous / current / next)| <---------+
+   *     |                |                           |
+   *     |                +---------------------------+
+   *     |                            |
+   *     |                            v
+   *     |            +--------------------------------------+     Y      +-------------------------+
+   *     |            |   finished finding adjacent iframes? +------------> exit and return adjacent|
+   *     |            +--------------------------------------+            |       array             |
+   *     |                            |                                   +-------------------------+
+   *     |                            |  N
+   *     |                            v
+   *     |                +---------------------------+
+   *     |                |                           |
+   *     |                | is cursor to the left of  |
+   *     |        Y       | storyIdx or IS storyIdx?  |       N
+   *     |          +-----+                           +------+
+   *     |          |     +---------------------------+      |
+   *     |          |                                        |
+   *     |          v                                        v
+   *     |  +---------------------------+    +----------------------+
+   *     |  |                           |    |                      |
+   *     |  |  move cursor to the right |    |  move cursor to the  |
+   *     |  |      of storyIdx          |    |  left of storyIdx    |
+   *     |  |                           |    |                      |
+   *     |  +---------------------------+    +----------------------+
+   *     |          |                                        |
+   *     |          |                                        |
+   *     |          +----------------------------------------+
+   *     |                               |
+   *     +-------------------------------+
+   *
    * @param {number} storyIdx
    * @param {number} maxIdx
-   * @return {!Array<number>}
+   * @return {!Array<!Object>}
    */
   findAdjacent(storyIdx, maxIdx) {
```

Comment:
I think there's a simpler algorithm here to be used, since this is really complex:
```js
const searchCenter = storyIdx + 0.2; // Prefer next iframes more than previous, this number can be between 0 and 0.5
const indexDistancesToCenter = stories.map((story) => {story, distance: Math.abs(story.index - searchCenter)});
const closestStories = indexDistancesToCenter.sort((a, b) => a.distance - b.distance).slice(MAX_IFRAMES_TO_LOAD)
```

WDYT?

Reply:
Yeah doing a sort would degrade performance to O(nlog(n)) while we are doing O(n) currently.  That algo also doesn't distinguish between the iframes that need to be positioned or loaded.

I simplified it more by adding two cursors and moving them further away from the currentIdx. Is that easier to understand?

Reply:
Much better 👍